Bencodex/Encoder.cs

```diff
@@ -286,27 +286,42 @@ long offset
             return 2L + lenStrLength + fingerprint.SerializeInto(buffer, offset);
         }
 
-        internal static long CountDecimalDigits(long value) => value < 10L
-            ? 1L
-            : value < 100L
-                ? 2L
-                : value < 1000L
-                    ? 3L
-                    : value < 10000L
-                        ? 4L
-                        : (long)Math.Floor(Math.Log10(value)) + 1L;
+        internal static long CountDecimalDigits(long value)
+        {
+#pragma warning disable SA1503 // Braces should not be omitted
+            if (value < 10L) return 1;
+            if (value < 100L) return 2;
+            if (value < 1000L) return 3;
+            if (value < 10000L) return 4;
+            if (value < 100000L) return 5;
+            if (value < 1000000L) return 6;
+            if (value < 10000000L) return 7;
+            if (value < 100000000L) return 8;
+            if (value < 1000000000L) return 9;
+            if (value < 10000000000L) return 10;
+            if (value < 100000000000L) return 11;
+            if (value < 1000000000000L) return 12;
+            if (value < 10000000000000L) return 13;
+            if (value < 100000000000000L) return 14;
+            if (value < 1000000000000000L) return 15;
+            if (value < 10000000000000000L) return 16;
+            if (value < 100000000000000000L) return 17;
+            if (value < 1000000000000000000L) return 18;
+            return 19;
+#pragma warning restore SA1503
+        }
 
         internal static long EncodeDigits(long positiveInt, byte[] buffer, long offset)
         {
-            string digits = positiveInt.ToString(CultureInfo.InvariantCulture);
-            if (offset < int.MaxValue)
+            const int asciiZero = 0x30; // '0'
+            long length = CountDecimalDigits(positiveInt);
+            for (long i = offset + length - 1; i >= offset; i--)
             {
-                return Encoding.ASCII.GetBytes(digits, 0, digits.Length, buffer, (int)offset);
+                buffer[i] = (byte)(positiveInt % 10 + asciiZero);
+                positiveInt /= 10;
```

Comment:
Could this formula be a little more helpful?🤔
```suggestion
                positiveInt = (int)(positiveInt * 0.1f);
```

Reply:
Nope, just tested it. It's even slower than the original. 😕

As for the rationale,
- Generally pure integer arithmetic should be faster
  - `ForLoopIntDiv` does no type conversion
  - `ForLoopFloatMult` goes through type changes twice (`long` -> `float` -> `long`)
- There's some danger of introducing floating point error

|           Method |     Mean |    Error |   StdDev |       Gen0 | Allocated |
|----------------- |---------:|---------:|---------:|-----------:|----------:|
|         GetBytes | 60.66 ms | 1.209 ms | 1.986 ms | 17777.7778 | 110.54 MB |
|    ForLoopIntDiv | 39.52 ms | 0.662 ms | 0.763 ms |          - |   3.81 MB |
| ForLoopFloatMult | 78.05 ms | 1.536 ms | 1.578 ms |          - |   3.82 MB |

Reply:
Oh, thanks to test it.👍